pkg/mediationcontainer/client_websocket_transport.go

```diff
@@ -145,8 +145,15 @@ func (wsTransport *ClientWebSocketTransport) closeAndResetWebSocket() {
 func (wsTransport *ClientWebSocketTransport) write(mtype int, payload []byte) error {
 	wsTransport.wsMux.Lock()
 	defer wsTransport.wsMux.Unlock()
-	wsTransport.ws.SetWriteDeadline(time.Now().Add(writeWaitTimeout))
-	return wsTransport.ws.WriteMessage(mtype, payload)
+
+	ws := wsTransport.ws
+	if ws == nil {
+		return errors.New("websocket connection unavailable")
+	}
+	if err := ws.SetWriteDeadline(time.Now().Add(writeWaitTimeout)); err != nil {
+		return err
+	}
```

Comment:
Since this function is shared by Ping, if we return error, the Ping goroutine will exit. If the websocket connection is reestablished, are we going to restart the Ping process?

Reply:
If I understand correctly, it looks like any error from the ping will not propagate up for the socket connection to be reset. Thus it will ping again until some other place detected the failure of the connection and reopen the socket connection.

Reply:
You can see this NPE occurs after we are trying to reconnect:
```
glog.V(2).Infof("[Reconnect] transport endpoint is closed, starting reconnect ...")

				// stop server messages listener
				remoteMediationClient.stopMessageHandler()
				// Reconnect
				err := transport.Connect(refreshTokenChannel, jwTokenChannel)
				// handle WebSocket creation errors
				if err != nil { //transport.ws == nil {
					glog.Errorf("[Reconnect] Initialization of remote mediation client failed: %v", err)
					close(disconnectFromTurbo)
					break
				}
```
We did try to build connection in wsTransport.performWebSocketConnection. Still since this whole process can not be atomic, anything can happen between.

Reply:
Or we can propagate the error in the ping all up and reset the websocket connection? if we do that, I am not sure what side effect of this will bring.

Reply:
I just want to confirm if the websocket is disconnected, and then later on reconnected, will the Ping be restarted? I have a feeling that it will **NOT**,  regardless of your fix.

Can we confirm? Just use the existing `kubeturbo` without your change, simply restart TP, and wait for `kubeturbo` to reconnect and confirm if Ping is actually restarted or not. If not, then this is a known issue, and we need to log a separate ticket and get it fixed.

Reply:
Not quite sure what you mean PIng restart.  From the log showing in the jira,  the NPE is thrown when doing the ping after this reconnecting message: [Reconnect] transport endpoint is closed, starting reconnect ...

Reply:
In the current implementation of startPing(), if an error occurs while sending the Ping message, the method logs an error message using the glog package and returns, which will cause the go wsTransport.startPing() goroutine to exit. Since no error handling is implemented for this goroutine, the error will be silently ignored and the Connect() method will continue to execute normally. So ping will restart upon websocket reconnect.  But its error will not cause websocket to reconnect.

Reply:
@libai98 , you are right, the Ping should be restarted as a part of the `Connect`.

I checked the code, I think there is a race condition going on. Here is the sequence of events I can think of:

1. Websocket gets disconnected
2. In `ListenForMessages()`, `wsTransport.ws.ReadMessage()` receives EOF (1006), and close the websocket
3. `wsTransport.stopListenerCh` is closed in `stopListenForMessages`
4. `wsTransport.connClosedNotificationCh` is notified

Now
* Step 3 should cause `startPing()` to exit, because it is selecting on `wsTransport.stopListenerCh`
* Step 4 should cause reconnect to happen

However, there is a race condition between `wsTransport.stopListenerCh` and `ticker.C` in `startPing()`, and when `ticker.C` arrives between step 2 and step 3, we get a segmentation fault because websocket is already closed.

Reply:
@ading1977 In this case, when we add error when ping fails as we did in this change. That will cause the startPing to exit.
	case <-ticker.C:
			glog.V(3).Infof("begin to send Ping message")
			if err := wsTransport.write(websocket.PingMessage, []byte{}); err != nil {
				glog.Errorf("Failed to send PingMessage to server:%v", err)
				return
			}